broadinstitute/single_cell_portal_core
# Cell filtering for violin plots (SCP-5275)

This extends cell filtering to violin plots, highlighting how expression distributions change upon subsetting.

### Overview
Previously (#1962), filtering only worked for cells in scatter plots.  That let users see how removing various combinations of cell annotations affected cluster scatter plots or similar gene expression scatter plots.  However, those visualizations are very dense.  So it was often difficult to discern certain important high-level changes by looking at scatter plots.

Now, with cell filtering applied to violin plots, it's much easier to discern how filtering out various cell annotations affects the _distribution_ of a given gene's expression within each group of a cell annotation.  For example, in JUN gene expression for the study "Cellular and transcriptional diversity over the course of human lactation" (SCP1671), filtering out cells with "Infant sick: no" noticeably affects expression distribution in neutrophils and other groups in the "General cell type" cell annotation.

Here's how it looks!

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/2ea88ca6-3958-404e-bb22-2aebf7e30993

### Next steps

This is intended as a minimally viable new feature.  Limitations and potential future work include:

1. Traces with no matching cells disappear, like cell filtering in scatter plots did initially.  It might help to keep these.
2. Distribution changes in some groups seem dramatic, but this may well be due to small sample size for the given group.  
3. The spinning icon still appears upon cell filtering.  Ideally, filtering would update things faster, and no loader would appear.  Enhancing the violin plot legend like the scatter plot legend would ameliorate that.
4. Animating distribution shifts upon filtering would aid object continuity, and perception of changes.

### Test
Automated tests have been updated to account for these changes.  To manually test:

1.  Go to study "Cellular and transcriptional diversity over the course of human lactation" ([SCP303](https://singlecell-staging.broadinstitute.org/single_cell/study/SCP303/cellular-and-transcriptional-diversity-over-the-course-of-human-lactation) on staging)
2. Search for gene "JUN"
3. Click "Distribution" tab
4. Deselect "Infant sick: no"
5. Confirm update occurs in traces showing distributions in each group of this cell annotation

This satisfies SCP-5275.
